File: Leetcode/MinimumPathSum.cpp
class Solution {
  using matrix = vector<vector<int>>;

 public:
  int minPathSum(const matrix& grid) {
    const auto n = grid.size(), m = grid[0].size();
    auto min_sum_to = matrix(n, vector<int>(m, -1));

    for (auto i = 0; i < n; ++i) {
      for (auto j = 0; j < m; ++j) {
        const auto option_a = i == 0 ? INT_MAX : min_sum_to[i - 1][j];
        const auto option_b = j == 0 ? INT_MAX : min_sum_to[i][j - 1];
        const auto better = min(option_a, option_b);

        min_sum_to[i][j] = (better == INT_MAX ? 0 : better) + grid[i][j];
      }
    }

    return min_sum_to.back().back();
  }
};

File: Leetcode/MinimumPathSum.rs
use std::cmp;

impl Solution {
    pub fn min_path_sum(grid: Vec<Vec<i32>>) -> i32 {
        let (n, m) = (grid.len(), grid[0].len());
        let mut min_sum_to = vec![vec![0; m]; n];
      
        let max_val = i32::max_value();
        for i in 0..n {
          for j in 0..m {
            let better_option = match (i, j) {
              (0, 0) => 0,
              (i, 0) => min_sum_to[i - 1][j],
              (0, j) => min_sum_to[i][j - 1],
              (i, j) => cmp::min(min_sum_to[i - 1][j], min_sum_to[i][j - 1]),
            };
            
            min_sum_to[i][j] = better_option + grid[i][j];
          }
        }

        return min_sum_to[n-1][m-1];
    }
}